test/integration/npm/routes.rollback.test.js

```diff
@@ -0,0 +1,177 @@
+      file: path.join(helpers.dirs.payloads, `${name}-${version}.json`)
+    }));
+    return macros.publishOk({
+      app: app,
+      registry: registry
+    }, {
+      file: path.join(helpers.dirs.payloads, `${name}-${version}.json`),
+      publishUrl: `http://localhost:8092/${name}`,
+      id: `${name}@${version}`,
+      path: `/${name}`
+    });
+  }
+
+  function createRelease(opts) {
+    cleanupWork.unshift(async.asyncify(app.release.delete.bind(app.release, opts)));
+    return async.asyncify(app.release.create.bind(app.release, opts));
+  }
+
+  function createDep(opts) {
+    cleanupWork.unshift(async.asyncify(app.release.dependent.remove.bind(app.release.dependent, opts)));
+    return async.asyncify(app.release.dependent.add.bind(app.release.dependent, opts));
+  }
+
+  function createBuild(opts) {
+    cleanupWork.unshift(app.bffs.unpublish.bind(app.bffs, opts));
+    return app.bffs.publish.bind(app.bffs, opts, publishOptions);
+  }
+
+  before(function (done) {
+    helpers.integrationSetup({
+      app: {
+        http: 8092,
+        npm: {
+          urls: {
+            read: 'http://localhost:8093',
+            write: {
+              default: 'http://localhost:8093'
+            }
+          }
+        }
+      },
+      registry: {
+        http: 8093
+      }
+    }, function (err, result) {
+      assume(err).to.be.falsey();
+      registry = result.registry;
+      app = result.app;
+      app.carpenter = app.publisher.carpenter = mocks.carpenter;
+
+      async.series([
+        publishOne({ name: 'parent-package', version: '0.0.1' }),
+        publishOne({ name: 'parent-package', version: '0.0.2' }),
+        publishOne({ name: 'parent-package', version: '0.0.3' }),
+        publishOne({ name: 'child-package', version: '0.0.1' }),
+        publishOne({ name: 'child-package', version: '0.0.2' }),
+        createRelease({ version: '0.0.1', pkg: 'parent-package' }),
+        createRelease({ version: '0.0.2', pkg: 'parent-package' }),
+        createRelease({ version: '0.0.1', pkg: 'child-package' }),
+        createRelease({ version: '0.0.2', pkg: 'child-package' }),
+        createDep({ pkg: 'parent-package', version: '0.0.1', dependent: 'child-package', dependentVersion: '0.0.1' }),
+        createDep({ pkg: 'parent-package', version: '0.0.2', dependent: 'child-package', dependentVersion: '0.0.2' }),
+        createBuild({ name: 'parent-package', version: '0.0.1', env: 'test' }, publishOptions),
+        createBuild({ name: 'parent-package', version: '0.0.2', env: 'test' }, publishOptions),
+        createBuild({ name: 'child-package', version: '0.0.1', env: 'test' }, publishOptions),
+        createBuild({ name: 'child-package', version: '0.0.2', env: 'test' }, publishOptions)
+      ], done);
+    });
+  });
+
+  after(function (done) {
+    async.series([
+      ...cleanupWork,
+      app.close.bind(app),
+      registry.close.bind(registry)
+    ], done);
+  });
+
+  // TODO: undo all that work in `after`
```

Comment:
Is this TODONE above?

Reply:
lol yeah. this was a note to myself as I was on my way home. 😂